### ISSUE TITLE
Be able to customize the global exception handler

### ISSUE THREAD

**Issue by LoicPoullain** [FoalTS/foal#370]
> This issue was previously named `Have hooks to catch rejected promises or uncaught errors`. See the latest comments to see the current state of the issue.

# Issue

In some situations, we want to catch errors thrown or promises rejected in a hook or a controller method. For example, during development, we may want to display a custom 500 page which prettifies the error stack.

# Possible solution

A possible solution could be to add a `CatcherPostHook` decorator which would take a `CatcherHookPostFunction` as parameter. The behavior would be analog to how errors, promises and express middlewares work.

*Empty example*
```typescript
@CatcherPostHook((err, ctx, services, response) => {

})
```

*Behavior demo*
```typescript
@Hook(() => {
  console.log(1);
  throw new Error('a');
})
@Hook(() => {
  console.log(2);
})
@CatcherPostHook((err, ctx, services, response) => {
  console.log(err.message);
  throw new Error('b');
})
@Hook(() => {
  console.log(3);
})
@CatcherPostHook((err, ctx, services, response) => {
  console.log(err.message);
})
@Hook(() => {
  console.log(4);
})

// Output:
// 1
// a
// b
// 4
```

**Comment by pdeak** [comment]
Just wondering, is there any workaround or hacky way to achieve a global exception handler across the app controller?

While prettifying error pages during development is nice, it's even more important for us that we can customize the default HTTP 500 page and add some additional behavior when such an unexpected error occurs.

**Comment by LoicPoullain** [comment]
Yes it is possible. One hacky solution to do it is to specify a "post" Express middleware to the `createApp` function in `src/index.ts`.

```typescript
// Example where Foal error handler is not used.
function errorHandler(err, req, res, next) {
  res.status(500).send(myPage);
}

// Example where Foal error handler is used.
function errorHandler(err, req, res, next) {
  // do something with err
  next(err);
}


const app = createApp(AppController, {
  postMiddlewares: [
    errorHandler
  ]
});
```

**Comment by LoicPoullain** [comment]
> add some additional behavior when such an unexpected error occurs in production.

Can you also give me an example of this? This will help me design something nice to handler properly errors in the future.

**Comment by pdeak** [comment]
@LoicPoullain The solution you posted above seems pretty good to me, and not even that hacky!

Basically, the two things we would have in a global error handler is to respond with a customized JSON payload, and to amend the logging of that error to indicate that it is in the context of the global catch-all handler.

**Comment by LoicPoullain** [comment]
Thanks!

**Comment by pdeak** [comment]
@LoicPoullain Actually, just a slight issue for us. We have to use a custom express instance for createApp, and the second parameter to createApp is expressInstanceOrOptions?: ExpressApplication | ExpressOptions. We would need a 3 parameter version of createApp which supports a custom express instance AND postMiddlewares.

**Comment by LoicPoullain** [comment]
Could you use preMiddlewares?

```typescript
const expressApp = express();
expressApp.use(middleware1);
expressApp.get('/a', middleware2);
const app = createApp(AppController, expressApp);

// is equivalent to

const app = createApp(AppController, {
  preMiddlewares: [
    middleware1
    express.Router().get('/a', middleware2)
  ],
  postMiddlewares: [ /*...*/ ]
});
```

If not, I can extend `ExpressOptions` this way (but you will have to wait until the end of the week or next week):
```typescript
const app = createApp(AppController, {
  expressIntance,
  postMiddlewares: [ /*...*/ ]
});
```

**Comment by pdeak** [comment]
Great thanks. I don't think preMiddlewares works because we want to catch errors in the chain exactly where the postMiddlewares is added: after all the other FoalTS middlewares, but before FoalTS's own notFound() and handleErrors() error handlers.

**Comment by LoicPoullain** [comment]
I'm not sure to understand your point here.

The architecture under the hood is pretty much like this:
```typescript
const app = express();
app.use(preMiddleware1);
app.use(preMiddleware2);
app.use(FoalTSMiddleware);
app.use(postMiddleware1);
app.use(postMiddleware2);
app.use(notFound());
app.use(handleErrors());
```

So error thrown in FoalTSMiddleware or preMiddleware1/2 are handled exactly the same

**Comment by pdeak** [comment]
But the global exception handler that we introduce needs to catch errors generated in FoalTSMiddleware too, so it has to be below it. It's main purpose is to catch all unexpected errors in our controllers, and the controllers are processed in FoalTSMiddleware presumably?

**Comment by LoicPoullain** [comment]
> But the global exception handler that we introduce needs to catch errors generated in FoalTSMiddleware too.

It does.

If you specify an error middleware (i.e. a function of this form `(err, req, res, next)=>{}`) in the `postMiddlewares` array, it will "catch" errors generated in your pre-middlewares as well as errors thrown/rejected in your hooks and controllers;

**Comment by pdeak** [comment]
Yep, postMiddleware is perfect for my use case. I was just stating that preMiddleware wouldn't work for a global error handler. Basically in my case I just need a postMiddleware and the ability to use a custom express instance.

**Comment by LoicPoullain** [comment]
Issue renamed to `Be able to customize the global exception handler`.

# Issue

As mentioned previously, the default global exception handler is not sufficient. Foal should have a dedicated solution to override it or complete it when needed.

Possible use cases:
- Send the exception to an external service (such as Sentry) with additional information (such as the connected user, the request path, etc).
- Customize the HTTP response returned to the client (a page, a JSON, etc).

# Solution

Add a new option to the `createApp` function.

```typescript
const app = createApp(AppController, {
  methods: {
    handleError: true,
  }
});
```

When enabled, we can add a method `handleError` in the `AppController` to handle errors.

```typescript
export class AppController {
  // ...
  async handleError(error: Error, details: ErrorDetails) {
    await this.sendToSentry(error);
    return renderError(error, details);
  }
}
```

The error details contains the `Context` object and will contain additional information in the future.

The `renderError` is the default function used to render the HTML page. This page varies depending on if the `debug` setting is enabled or not.

**Comment by LoicPoullain** [comment]
The PR has been merged. The feature will be released with v1.4.0.
